docs/uvm_guide.md

```diff
@@ -0,0 +1,99 @@
+
+### Analysis Components
+#### Coverage Collector
+The coverage collector is used to collect functional coverage information so that one can better know what they need to write tests for, or conversely, what they have already tested.  This coverage collector needs to asynchronously handle coverage information for all four interfaces of the DUT.  Therefore, we use a UVM construct called analysis imps to define an analysis port for each interface's monitor to broadcast to.  Then, every time that the coverage collector sees a transaction on its analysis port, it will sample coverage information for that interface with the given transaction.  The coverage information is defined in coverage groups in the coverage collector, with specific bins defined where necessary.  Coverage groups, coverage points, and bins are all part of the SystemVerilog standard, so look to the IEEE specification for more information.
+
+#### Scoreboard
+In the UVM the scoreboard is used to compare the output of the DUT against some emulated/functional model of the DUT to verify that the output is what is expected.  Herein, our scoreboard has three analysis ports that receive transactions from the input monitor, the tlb monitor, and the output monitor.  These three ports are implemented as analysis imps so that when each receives a transaction, it handles it accordingly.  The input and tlb transactions are sent to the predictor as they are received.  In addition, the input transaction is queued in a 2 item long queue, such that the cache request lines up with the cache response from the output monitor.  When a transaction comes in on the output port, we first pop an item from the input queue so that it remains 2 items long.  Then check if the transaction is valid and if it is we assign an id to the transaction (used by the comparator) which is the virtual address of the input transaction.  Then we send that output transaction to the comparator.
+
+#### Predictor
+In the UVM the predictor is a functional model of the cache which should give the same output as the actual cache for the sake of comparison.  In its current state, the predictor written here is not a complex functional model, but instead is hardcoded to be accurate for the test_load and test_uncached_load tests.  This is a place for future work, which we greatly encourage someone to come along and complete.  Currently, the predictor takes the input and tlb transaction from its input and if they are valid, gives an output that is valid and indexed according to the input transaction virtual address.  This output is broadcast to the comparator to be compared to the output from the scoreboard (and inherently, the DUT).
```

Comment:
what does hardcoded mean in this context?

Reply:
As it stands currently one of the big things that need to be added to improve this testbench is a functional cache model to predict the correct output of the cache.  At the moment it is "hardcoded" in that the predictor just predicts the cache to respond with the value of the virtual address as this is what is stored in prog.mem currently.  Were we to change what is stored in prog.mem this would no longer work.

Reply:
I see. Could we not intake prog.mem as use that as the predictor source of truth?